Generalize the Mangling of Constrained Existential Types

CodaFi opened the pull request:
Upgrade the old mangling from a list of argument types to a
list of requiremnets. For now, only same-type requirements
may actually be mangled since those are all that are available
to the surface language.

Reconstruction of existential types now consists of demangling (a list of)
base protocol(s), decoding the constraints, and converting the same-type
constraints back into a list of arguments. Which is simple to do at the moment
because the only kinds of same-type constraints we have are rooted at
dependent types of `Self`.

rdar://96088707

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

rjmccall commented on docs/ABI/Mangling.rst:
```diff
@@ -638,7 +638,7 @@ Types
   type ::= protocol-list superclass 'Xc'     // existential type with superclass
   type ::= protocol-list 'Xl'                // existential type with AnyObject
-  type ::= protocol-list 'y' (type* '_')* type* retroactive-conformance* 'XP'   // parameterized protocol type
+  type ::= protocol-list 'y' (requirement* '_')* requirement* 'XP'   // constrained existential type
```
Why are there mulltiple levels of requirements here?

CodaFi replied:
Thinko, corrected.

rjmccall commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1233,7 +1233,12 @@ void ASTMangler::appendType(Type type, GenericSignature sig,
-      appendType(EMT->getInstanceType(), sig, forDecl);
+      if (EMT->getInstanceType()->isExistentialType() &&
+          EMT->hasParameterizedExistential())
+        appendConstrainedExistential(EMT->getInstanceType(), sig, forDecl);
```
Metatypes can be nested

CodaFi replied:
`EMT->getInstanceType()->isExistentialType()` ensures that we don't descend into them. `ExistentialMetatype` is only considered by `isAnyExistentialType()`.

rjmccall replied:
Okay, so if we don't need to descend into nested metatypes here, why do we need to call out direct existential metatypes at all instead of just letting the recursive visitation in `appendType` handle it?

rjmccall replied:
Like, my patch to symbolically reference shapes does need to do something at this level because the existential metatype is part of the shape, but I think the normal mangling path doesn't have to.  You'll just get down to the `ExistentialType` and do what you need to do at that level.

CodaFi replied:
> why do we need to call out direct existential metatypes at all instead of just letting the recursive visitation in appendType handle it?

The AST for `any P<T>.Type` currently looks like `(ExistentialMetatypeType (ParameterizedProtocolType ...))` which is unfortunate...

rjmccall commented on test/Demangle/Inputs/manglings.txt:
```diff
@@ -428,9 +428,8 @@ $s1A3bar1aySSYt_tF ---> A.bar(a: _const Swift.String) -> ()
-$s4test3FooVAAyyAA1P_pyxqd__XPlF ---> test.Foo.test<A>(test.P<A, A1>) -> ()
-$s4test3FooVAAyyAA1P_pyxXPF ---> test.Foo.test(test.P<A>) -> ()
+$s4test3fooyyAA1P_px1TRtz_XPlF ---> test.foo<A>(any test.P<A.T == A>) -> ()
+$s4test3fooyyAA1P_pSS1TAaCPRtz_Si1UAERtzXPF ---> test.foo(any test.P<A.test.P.T == Swift.String, A.test.P.U == Swift.Int>) -> ()
```
Oof, this demangling is super confusing.  I guess we're in a generic context that says τ_0_0 is such-and-such type?  Maybe we can push a different context when the LHS of these?

This ties in with my concern about using requirements here — it's really a false economy, because τ_0_0 means different things in different positions in the requirement.  And you still don't have the ability to express things like `P<.Slice == .self>`, because you can't write τ_0_0 on the RHS without it being ambiguous in case you're actually in a generic context.

rjmccall replied:
If you want to stick with mangling requirements, you could just build them in terms of some new "protocol self" base type that's guaranteed to mangle differently from other types.  Or you could use `Requirement` internally in the compiler when building + uniquing things and then do a special-case mangling at the last minute.

CodaFi committed:
Generalize the Mangling of Constrained Existential Types

Upgrade the old mangling from a list of argument types to a
list of requiremnets. For now, only same-type requirements
may actually be mangled since those are all that are available
to the surface language.

Reconstruction of existential types now consists of demangling (a list of)
base protocol(s), decoding the constraints, and converting the same-type
constraints back into a list of arguments.

rdar://96088707

rjmccall commented on lib/AST/ASTMangler.cpp:
```diff
@@ -2797,7 +2803,8 @@ bool ASTMangler::appendGenericSignature(GenericSignature sig,
 void ASTMangler::appendRequirement(const Requirement &reqt,
-                                   GenericSignature sig) {
+                                   GenericSignature sig,
+                                   bool baseIsProtocolSelf) {
```
Could you name this in some way that makes it clear it only applies to the first type?

CodaFi replied:
Named and documented.

rjmccall commented:
Robert, you should not be merging patches like this with just smoke tests.
